File: app/src/main/java/com/websarva/wings/android/zuboradiary/di/RepositoryModule.java
package com.websarva.wings.android.zuboradiary.di;

import androidx.annotation.NonNull;
import androidx.work.WorkManager;

import com.websarva.wings.android.zuboradiary.data.database.DiaryDAO;
import com.websarva.wings.android.zuboradiary.data.database.DiaryDatabase;
import com.websarva.wings.android.zuboradiary.data.database.DiaryItemTitleSelectionHistoryRepository;
import com.websarva.wings.android.zuboradiary.data.database.DiaryRepository;
import com.websarva.wings.android.zuboradiary.data.database.DiaryItemTitleSelectionHistoryDAO;
import com.websarva.wings.android.zuboradiary.data.network.WeatherApiRepository;
import com.websarva.wings.android.zuboradiary.data.network.WeatherApiResponse;
import com.websarva.wings.android.zuboradiary.data.network.WeatherApiService;
import com.websarva.wings.android.zuboradiary.data.preferences.SettingsRepository;
import com.websarva.wings.android.zuboradiary.data.preferences.UserPreferences;
import com.websarva.wings.android.zuboradiary.data.worker.WorkerRepository;

import java.util.Objects;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import dagger.hilt.InstallIn;
import dagger.hilt.components.SingletonComponent;

@Module
@InstallIn(SingletonComponent.class)
public class RepositoryModule {
    @Singleton
    @Provides
    @NonNull
    public static DiaryRepository provideDiaryRepository(
            DiaryDatabase diaryDatabase,
            DiaryDAO diaryDAO,
            DiaryItemTitleSelectionHistoryDAO diaryItemTitleSelectionHistoryDAO) {
        Objects.requireNonNull(diaryDatabase);
        Objects.requireNonNull(diaryDAO);
        Objects.requireNonNull(diaryItemTitleSelectionHistoryDAO);

        DiaryRepository repository =
                new DiaryRepository(diaryDatabase, diaryDAO, diaryItemTitleSelectionHistoryDAO);
        return Objects.requireNonNull(repository);
    }

    @Singleton
    @Provides
    @NonNull
    public static DiaryItemTitleSelectionHistoryRepository provideEditDiarySelectItemTitleRepository(
            DiaryItemTitleSelectionHistoryDAO diaryItemTitleSelectionHistoryDAO) {
        Objects.requireNonNull(diaryItemTitleSelectionHistoryDAO);

        DiaryItemTitleSelectionHistoryRepository repository =
                new DiaryItemTitleSelectionHistoryRepository(diaryItemTitleSelectionHistoryDAO);
        return Objects.requireNonNull(repository);
    }

    @Singleton
    @Provides
    @NonNull
    public static SettingsRepository provideSettingsRepository(UserPreferences userPreferences) {
        Objects.requireNonNull(userPreferences);

        SettingsRepository repository = new SettingsRepository(userPreferences);
        return Objects.requireNonNull(repository);
    }

    @Singleton
    @Provides
    @NonNull
    public static WorkerRepository provideWorkerRepository(WorkManager workManager) {
        Objects.requireNonNull(workManager);

        WorkerRepository repository = new WorkerRepository(workManager);
        return Objects.requireNonNull(repository);
    }

    @Singleton
    @Provides
    @NonNull
    public static WeatherApiRepository provideWeatherApiRepository(WeatherApiService weatherApiService) {
        Objects.requireNonNull(weatherApiService);

        WeatherApiRepository repository = new WeatherApiRepository(weatherApiService);
        return Objects.requireNonNull(repository);
    }
}

File: app/src/main/java/com/websarva/wings/android/zuboradiary/di/RepositoryModule.kt
package com.websarva.wings.android.zuboradiary.di

import com.websarva.wings.android.zuboradiary.data.database.DiaryDataSource
import com.websarva.wings.android.zuboradiary.data.location.FusedLocationDataSource
import com.websarva.wings.android.zuboradiary.data.network.WeatherApiDataSource
import com.websarva.wings.android.zuboradiary.data.repository.DiaryRepository
import com.websarva.wings.android.zuboradiary.data.repository.WeatherInfoRepository
import com.websarva.wings.android.zuboradiary.data.repository.UserPreferencesRepository
import com.websarva.wings.android.zuboradiary.data.preferences.UserPreferences
import com.websarva.wings.android.zuboradiary.data.repository.UriRepository
import com.websarva.wings.android.zuboradiary.data.repository.WorkerRepository
import com.websarva.wings.android.zuboradiary.data.uri.UriPermissionDataSource
import com.websarva.wings.android.zuboradiary.data.worker.ReminderNotificationWorkManager
import dagger.Module
import dagger.Provides
import dagger.hilt.InstallIn
import dagger.hilt.components.SingletonComponent
import javax.inject.Singleton

@Module
@InstallIn(SingletonComponent::class)
internal object RepositoryModule {

    @Singleton
    @Provides
    fun provideDiaryRepository(
        diaryDataSource: DiaryDataSource
    ): DiaryRepository = DiaryRepository(diaryDataSource)

    @Singleton
    @Provides
    fun provideUserPreferencesRepository(
        userPreferences: UserPreferences
    ): UserPreferencesRepository = UserPreferencesRepository(userPreferences)

    @Singleton
    @Provides
    fun provideWorkerRepository(
        workManager: ReminderNotificationWorkManager
    ): WorkerRepository = WorkerRepository(workManager)

    @Singleton
    @Provides
    fun provideWeatherApiRepository(
        weatherApiDataSource: WeatherApiDataSource,
        fusedLocationDataSource: FusedLocationDataSource
    ): WeatherInfoRepository =
        WeatherInfoRepository(
            weatherApiDataSource,
            fusedLocationDataSource
        )

    @Singleton
    @Provides
    fun provideUriRepository(
        uriPermissionDataSource: UriPermissionDataSource
    ): UriRepository = UriRepository(uriPermissionDataSource)
}
